### PR TITLE
GPU support - September 2024

### PR DESCRIPTION
**Pull request by TeachRaccooon**
This PR inherits commits originally introduced in PR #67.
The discussion of some of the details can also be found there.
The list of changes is as follows:
1. Introduces a CMake build option for GPU support (specifically, CUDA support) in RandLAPACK. This is enabled with ``-DRequireCUDA=ON``.
2. Introduces rl_cuda_kernels.cuh - file contains various utility GPU functions, including some BLAS and LAPACK-level routines. 
3. Introduces rl_cqrrpt_gpu.cuh, - a GPU version of CQRRPT. Note that since many parts of CQRRPT (including sketching) do not (currently) have GPU versions, the data offload happens inside of the algorithm. The input data is expected to be located on a CPU.
4. Introduces rl_cqrrp_gpu.cuh - a GPU version of CQRRP algorithm, which accepts data allocated on a GPU. 
5. Includes tests for the functions from the above files and benchmarks (living in test space) for CQRRP algorithm. In the future, these should be moved into benchmarking space (built separately). For now, we can avoid running these with the rest of the tests by using `ctest --gtest_filter=-*bench*`.

Issues #77 - #80 are related to this PR.

### CODE REVIEW

**Comment by TeachRaccooon** [comment]
@rileyjmurray I moved commits from PR #67 here.
Still need to perform a cleanup and make sure everything works with no GPU available.

**Comment by TeachRaccooon** [comment]
@rileyjmurray I've included most of the fixes you requested (and opened issues #77 - #80 for the ones that require nontrivial effort) and added the optional GPU support. 
To configure the project with GPU support, using the "-DRequireCUDA=ON" flag is required.
Lmk if anything else needs to be updated.

File: RandLAPACK.hh [hunk 1]
```diff
@@ -23,4 +23,11 @@
 #include "RandLAPACK/drivers/rl_revd2.hh"
 #include "RandLAPACK/drivers/rl_rbki.hh"
 
+// Cuda functions - issues with linking/visibility when present here.
+// Added a temporary fix to pass tests, will hopefully not encounter the issue of
+// function repitition in benchmarks.
+//#include "RandLAPACK/drivers/rl_cqrrpt_gpu.hh"
+//#include "RandLAPACK/drivers/rl_cqrrp_gpu.hh"
+//#include "RandLAPACK/gpu_functions/rl_cuda_kernels.cuh"
+
```
Comment (rileyjmurray) [inline RandLAPACK.hh:32]: Please remove these comments or explain why you still need them.

File: RandLAPACK/gpu_functions/rl_cusolver.hh [hunk 2]
```diff
@@ -0,0 +1,61 @@
+#ifndef bpcg_cublas_h
+#define bpcg_cublas_h
```
Comment (rileyjmurray) [inline RandLAPACK/gpu_functions/rl_cusolver.hh:2]: Use ``#pragma once``.

Please remove the explicit include guard. If you have to keep it, please give the macro a better name.

**Review by rileyjmurray — CHANGES_REQUESTED** [review]
Looks good. I have a couple minor comments, and one comment that'll make you edit a ton of files:

Most files have include guards of the form
```c++
#pragma once

#ifndef randlapack_cqrrpt_h
#define randlapack_cqrrpt_h
...
#endif
```
These guards are redundant. One you have pragma once you can remove the ``#ifndef ... #endif`` macros. Please do so in this PR. If you want to start by only removing them for .hh files (as opposed to .cuh files) that'd be okay with me.

File: RandLAPACK/gpu_functions/rl_cusolver.hh (hunk 2 above)
Comment (TeachRaccooon) [reply RandLAPACK/gpu_functions/rl_cusolver.hh:2]: Done.

**Comment by TeachRaccooon** [comment]
> Looks good. I have a couple minor comments, and one comment that'll make you edit a ton of files:
> 
> Most files have include guards of the form
> 
> ```c++
> #pragma once
> 
> #ifndef randlapack_cqrrpt_h
> #define randlapack_cqrrpt_h
> ...
> #endif
> ```
> 
> These guards are redundant. One you have pragma once you can remove the `#ifndef ... #endif` macros. Please do so in this PR. If you want to start by only removing them for .hh files (as opposed to .cuh files) that'd be okay with me.

@rileyjmurray Addressed it for all .hh and .cuh files.

**Review by rileyjmurray — APPROVED** [review]
You may press the Big Green Button!

**Pull request merged**